### ISSUE TITLE
Pure environment model for binstar tokens

### ISSUE THREAD

**Issue by byjrack** [conda/conda#14228]
### Checklist

- [X] I added a descriptive title
- [X] I searched open requests and couldn't find a duplicate

### What is the idea?

https://github.com/conda/conda/blob/main/conda/gateways/anaconda_client.py#L31

So read_binstar seems to only allow for filesystem based tokens.  if using the commercial .cloud repository that means that in an image build the token gets captured in the filesystem even when using secrets mounts in `docker build`.  I am having folks do a `conda token remove` at the end, but there is a filesystem layer where that is caught.

Be nice if that function also allowed for pulling from a predefined env var and using.  Might need a nudge earlier in the condarc as well, but helps to really avoid secret material from getting caught in the artifacts along the way.

If there is a method to use binstar that doesn't result in a layer capturing the secret i am all ears.

### Why is this needed?

If conda is used with anaconda.cloud the binstar token gets caught up in the filesystem and can be stuck in a docker image unintentionally.

### What should happen?

Enable read_binstar to pull from an env and not just the filesystem so token is completely ephemeral and can follow the standard docker secret handling.

**Comment by travishathaway** [comment]
I think setting tokens/secrets as environment variable would be a nice feature addition. The thing that complicates this is that tokens/secrets are specific per channel, so we would have to figure out how this would be specified with the environment variables.

I do think that this is a problem better tackled in a project such as [conda-auth](https://github.com/conda-incubator/conda-auth) though. In fact, because we have the [auth handlers plugin hook](https://docs.conda.io/projects/conda/en/stable/dev-guide/plugins/auth_handlers.html) available, anyone could come up with a solution for this by writing their own conda plugin.

**Comment by byjrack** [comment]
Is the access to anaconda.cloud moving from `conda token` to `conda auth` sometime soon?  That seemed focused on basic auth instead of a bearer token so by default it was looking at a different concern, but standardizing on a single auth framework does seem like the right way to go.

I had tried to open the Issue to the [token repo](https://github.com/anaconda/conda-token), but that looks to be private so ended up here in the core package manager as some of the anaconda binstar stuff seemed pretty hard coded.  

And good point on the N private channel support as that gets more complicated.  Hadn't really thought of that as I only deal with the Anaconda commercial registry for the time being.  Not sure if the Git style http_proxy global assumption vs the scoped in gitconfig would work for a majority of cases that need to go env vs file system.

Let me toy around with an Auth plugin.  Basically `add_anaconda_token` to false so it stays out of the classic binstar path and replicate the existing `token` functionality sourcing from the env.

**Comment by byjrack** [comment]
Yeah going to take a bit more than the simple authplugin on the surface.  Seems like a pre-command might be able to intercept `install`,`upgrade`,etc and manipulate the base, but feels super fragile.

Looks like the Anaconda commercial repo uses a site/t/<token>/repo/... path for auth rather than a token@site/repo/... or some sort of Auth Header.  Basically binstar is sort of a special case compared to a standard www-authenticate style authentication.  Checking with some folks on that side though as maybe there are less documented paths.

Assuming I can get a plugin that does the auth if we use a distribution like miniconda how would i pull that custom plugin into conda's entrypoint namespace?  I get the channel_settings, but missing something obvious on how conda would see this plugin.  Would it be a full package so do an install of custom-auth-plugin from my local and then when conda hits the channel_settings config it would know to follow that path?
